src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java

```diff
@@ -28,13 +73,145 @@ private static AnswerType getMatchingResultFromMatchingMethodsList(String contac
+    }
+
+    private static boolean isMatchWithHyphenatedNames(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
+        // Also checks names if they have more than one hyphen in them
+        // Splits the name(s) into arrays, and runs a compare check.
+        // If any match, returns true (for every compare check).
+        // If both first and last name compare checks return true, method returns true.
+
+        ArrayList<String> contactOneLastNamesWithoutHyphens = seperateHypenNames(contactOneFirstName);
+        ArrayList<String> contactOneFirstNamesWithoutHyphens = seperateHypenNames(contactOneLastName);
+        ArrayList<String> contactTwoFirstNamesWithoutHyphens = seperateHypenNames(contactTwoFirstName);
+        ArrayList<String> contactTwoLastNamesWithoutHyphens = seperateHypenNames(contactTwoLastName);
+        boolean firstNamesMatch = isNameInGroupNamesMatch(contactOneLastNamesWithoutHyphens, contactTwoFirstNamesWithoutHyphens);
+        boolean lastNamesMatch = isNameInGroupNamesMatch(contactOneFirstNamesWithoutHyphens, contactTwoLastNamesWithoutHyphens);
+
+        return (firstNamesMatch & lastNamesMatch);
+    }
+
+    private static boolean isFirstNameAndLastNameSwitched(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
+        return (contactOneFirstName.equals(contactTwoLastName) && contactOneLastName.equals(contactTwoFirstName));
+    }
+
+    private static boolean isInitialMatchNameWithoutPrefix(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
+        String contactOneFirstNameWithoutPunctuations = removePunctuationsInName(contactOneFirstName);
+        String contactOneLastNameWithoutPunctuations = removePunctuationsInName(contactOneLastName);
+        String contactTwoFirstNameWithoutPunctuations = removePunctuationsInName(contactTwoFirstName);
+        String contactTwoLastNameWithoutPunctuations = removePunctuationsInName(contactTwoLastName);
+        boolean firstNameIsInitialOf2ndFirstName = false;
+        boolean lastNameIsInitialOf2ndLastName = false;
+        boolean firstNamesMatchFound = false;
+        boolean lastNamesMatchFound = false;
+
+        // First name is an initial
+        if (isNameInitial(contactOneFirstNameWithoutPunctuations)){
+            // Initial matches the first name
+            if (contactTwoFirstNameWithoutPunctuations.startsWith(contactOneFirstNameWithoutPunctuations)){
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        } else if (isNameInitial(contactTwoFirstNameWithoutPunctuations)){
+            if (contactOneFirstNameWithoutPunctuations.startsWith(contactTwoFirstNameWithoutPunctuations)){
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        }
+
+        // Last name is an initial
```

Comment:
Last name can never match if it is just an initial. I don't even think we can call that one a maybe. We should add a test to ensure that we never declare a name match in this case as I think it would result in a bunch of bad recommendations.

Reply:
I'm going to give a case here: When I was growing up with someone, he always was called by his initials TJ. So in two database he might be known as (firstName: T | lastName: J) and (firstname: Tom | lastName: Jacobs).

That being said, confirming you don't think the match should be a maybe, because the last name is an initial?

Reply:
If it is an exact match on the last name, we should have it match. But I wouldn't do the "does the first letter match the initial" on last name. For example, where C1 is Contact1 and C2 is Contact2, and F is First, L is Last.... C1F: T; C1L: J should have a "maybe match" on C2F:Tommy, C2L:J, a "match" on C2F:T, C2L:J, and a "no match" on C2F:T, C2L:Johnson. We can discuss more in the call but if you agree, can you add these tests before making the change?